File: src/layers/auth-layer/python/auth_jwt.py
import uuid

import jwt


def encode(
    key: str,
    type: str,
    exp: int,
    id="",
    sub="",
    name="",
    aud="",
    dest="",
    roles: list[str] = [],
    orig="",
) -> str:
    payload = {"typ": type, "jti": id or str(uuid.uuid4()), "exp": exp}

    if sub:
        payload["sub"] = sub

    if name:
        payload["name"] = name

    if aud:
        payload["aud"] = aud

    if dest:
        payload["dest"] = dest

    if roles:
        payload["roles"] = roles

    if orig:
        payload["orig"] = orig

    return jwt.encode(payload=payload, key=key, algorithm="HS512")

File: src/auth-lib/src/auth_jwt.rs
use crate::auth_enums::{Action, Role};
use serde::{Deserialize, Serialize};

#[derive(
    Clone, Copy, Debug, Default, Hash, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize,
)]
#[serde(rename_all = "snake_case")]
pub enum SessionTokenType {
    #[default]
    Session,
}

#[derive(Debug, PartialEq, Serialize, Deserialize)]
pub struct AuthSessionToken {
    #[serde(default)]
    pub typ: SessionTokenType,

    pub jti: String,
    pub sub: u32,
    pub exp: u64,
    pub aud: Action,
    pub dest: Action,
}

#[derive(
    Clone, Copy, Debug, Default, Hash, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize,
)]
#[serde(rename_all = "snake_case")]
pub enum RefreshTokenType {
    #[default]
    Refresh,
}

#[derive(Debug, PartialEq, Serialize, Deserialize)]
pub struct AuthRefreshToken {
    #[serde(default)]
    pub typ: RefreshTokenType,

    pub jti: String,
    pub sub: u32,
    pub exp: u64,
    pub aud: Action,
}

#[derive(
    Clone, Copy, Debug, Default, Hash, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize,
)]
#[serde(rename_all = "snake_case")]
pub enum AccessTokenType {
    #[default]
    Access,
}

#[derive(Debug, PartialEq, Serialize, Deserialize)]
pub struct AuthAccessToken {
    #[serde(default)]
    pub typ: AccessTokenType,

    pub jti: String,
    pub sub: u32,
    pub exp: u64,
    pub roles: Vec<Role>,
    pub orig: String,
}
